pkg/services/ngalert/state/state_tracker.go

```diff
@@ -24,77 +31,132 @@ type cache struct {
+	return st.stateCache.cacheMap[stateId]
 }
 
-func (st *StateTracker) ProcessEvalResults(uid string, results eval.Results, condition models.Condition) []AlertState {
+func (st *StateTracker) ProcessEvalResults(uid string, results eval.Results, condition ngModels.Condition) []AlertState {
+	st.Log.Info("state tracker processing evaluation results", "uid", uid, "resultCount", len(results))
 	var changedStates []AlertState
 	for _, result := range results {
-		currentState := st.stateCache.getOrCreate(uid, result)
-		currentState.Results = append(currentState.Results, result.State)
-		newState := st.getNextState(uid, result)
-		if newState != currentState.State {
-			currentState.State = newState
-			changedStates = append(changedStates, currentState)
+		if s, ok := st.setNextState(uid, result); ok {
+			changedStates = append(changedStates, s)
 		}
-		st.stateCache.update(currentState)
 	}
+	st.Log.Debug("returning changed states to scheduler", "count", len(changedStates))
 	return changedStates
 }
 
-func (st *StateTracker) getNextState(uid string, result eval.Result) eval.State {
-	currentState := st.stateCache.getOrCreate(uid, result)
-	if currentState.State == result.State {
-		return currentState.State
-	}
-
+//Set the current state based on evaluation results
+//return the state and a bool indicating whether a state transition occurred
+func (st *StateTracker) setNextState(uid string, result eval.Result) (AlertState, bool) {
+	currentState := st.getOrCreate(uid, result)
+	st.Log.Debug("setting alert state", "uid", uid)
 	switch {
 	case currentState.State == result.State:
-		return currentState.State
+		st.Log.Debug("no state transition", "cacheId", currentState.CacheId, "state", currentState.State.String())
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, false
 	case currentState.State == eval.Normal && result.State == eval.Alerting:
-		return eval.Alerting
+		st.Log.Debug("state transition from normal to alerting", "cacheId", currentState.CacheId)
+		currentState.State = eval.Alerting
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.StartsAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.EndsAt = strfmt.DateTime(result.EvaluatedAt.Add(40 * time.Second))
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, true
 	case currentState.State == eval.Alerting && result.State == eval.Normal:
-		return eval.Normal
+		st.Log.Debug("state transition from alerting to normal", "cacheId", currentState.CacheId)
+		currentState.State = eval.Normal
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.EndsAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, true
 	default:
```

Comment:
This might be a gap in my knowledge of how we expect alerting to function at play here. I would expect that an evaluation result would eventually return to normal, once the condition is no longer being met/violated.